src/main/java/com/vmware/vipclient/i18n/messages/api/opt/local/LocalLocaleOpt.java

```diff
@@ -37,72 +39,104 @@ public LocalLocaleOpt(LocaleDTO dto) {
 		}
-		languagesNames = getLanguagesNamesFromBundle(locale);
-		if (languagesNames != null) {
+		languagesNames = new LocaleCacheItem();
+		getLanguagesNamesFromBundle(locale, languagesNames);
+		if (!languagesNames.getCachedData().isEmpty()) {
 			logger.debug("Find the languages' names from local bundle for locale [{}].\n", locale);// [datetime] and
 			formattingCacheService.addLanguagesNames(locale, languagesNames);
 			logger.debug("Languages' names are cached for locale [{}]!\n\n", locale);
-			return languagesNames;
 		}
-		return null;
+		return languagesNames.getCachedData();
 	}
 
-	private JSONObject getLanguagesNamesFromBundle(String locale) {
+	private void getLanguagesNamesFromBundle(String locale, LocaleCacheItem cacheItem) {
 		logger.debug("Look for languages' names from local package bundle for locale [{}]!", locale);
 		String normalizedLocale = CommonUtil.getCLDRLocale(locale, localePathMap, localeAliasesMap);
 		logger.info("Normalized locale for locale [{}] is [{}]", locale, normalizedLocale);
 		if(normalizedLocale == null || normalizedLocale.isEmpty())
-			return null;
-        try {
-            String languagesJsonStr = PatternUtil.getLanguageFromLib(normalizedLocale);
-			JSONObject languagesData = (JSONObject) new JSONParser().parse(languagesJsonStr);
-            return (JSONObject) languagesData.get(PatternKeys.LANGUAGES);
+			return;
+		try {
+			String languagesJsonStr = PatternUtil.getLanguageFromLib(normalizedLocale);
+			Map<String, Object> languagesData = (Map<String, Object>) new JSONParser().parse(languagesJsonStr);
+			if (languagesData != null) {
+				cacheItem.addCachedData((Map<String, String>) languagesData.get(PatternKeys.LANGUAGES));
+				cacheItem.setTimestamp(System.currentTimeMillis());
+			}
 		} catch (Exception e) {
 			e.printStackTrace();
 			logger.error(e.getMessage());
 		}
-        return null;
 	}
 
-	public JSONObject getRegions(String locale) {
+	public void getRegions(String locale, LocaleCacheItem cacheItem) {
 		logger.debug("Look for regions from local bundle for locale [{}]!", locale);
 		String normalizedLocale = CommonUtil.getCLDRLocale(locale, localePathMap, localeAliasesMap);
 		logger.info("Normalized locale for locale [{}] is [{}]", locale, normalizedLocale);
 		if(normalizedLocale == null || normalizedLocale.isEmpty())
-			return null;
-        try {
-		    String regionsJsonStr = PatternUtil.getRegionFromLib(normalizedLocale);
-			JSONObject regionsData = (JSONObject) new JSONParser().parse(regionsJsonStr);
-            return (JSONObject) regionsData.get(PatternKeys.TERRITORIES);
+			return;
+		try {
+			String regionsJsonStr = PatternUtil.getRegionFromLib(normalizedLocale);
+			Map<String, Object> regionsData = (Map<String, Object>) new JSONParser().parse(regionsJsonStr);
+			if (regionsData != null) {
+				cacheItem.addCachedData((Map<String, String>) regionsData.get(PatternKeys.TERRITORIES));
+				cacheItem.setTimestamp(System.currentTimeMillis());
```

Comment:
To avoid mismatched etag, timestamp and maxAgeMillis due to thread unsafety, have to set all the factors that go together (etag, timestamp and maxAgeMillis, and cachedData) in one synchronized setter method.
So can't move set timestamp to service layer.